New devices: Blackberry KEY2 (athena) and KEY2 LE (luna)

sajattack opened the pull request:
Adds: 
- panel driver
- keypad driver
- dts

Updates:
- sdm660_defconfig

[dmesg.txt](https://github.com/user-attachments/files/21272461/dmesg.txt)

-------------------------------------------------
### unmerged 3rdparty patches
- stmpe1801: https://patchwork.kernel.org/project/linux-input/patch/20210618145149.10136-1-cerg2010cerg2010@mail.ru/
-------------------------------------------------

cerg2010cerg2010 committed:
Input: stmpe-keypad - add STMPE1801 support

This chip variant differs from others - it has different register
map and swapped rows/columns.

Signed-off-by: Sergey Larin <cerg2010cerg2010@mail.ru>

sajattack committed:
dts: qcom: sdm660: Add sdm660-bbry-common, sdm660-bbry-athena-common, sdm636-bbry-luna-common, sdm660-bbry-athena-boe, and sdm636-bbry-luna-boe

for BlackBerry KEY2 and KEY2 LE (boe/livata panel variant)

working features:
- Internal MMC
- External SDcard
- Wi-Fi
- Bluetooth
- Keypad
- Panel
- GPU (bit crashy)
- fuel gauge
- volume buttons
- usb

TODO:
- keypad backlight
- modem
- audio
- 3-button row
- gps
- fm radio
- camera
- synaptics panel variants

Signed-off-by: Paul Sajna <hello@paulsajna.com>

sajattack committed:
drm: panel: Add boe,bv045fhm-l00

Used by some BlackBerry phones including KEY2 LE (luna)

https://www.panelook.com/BV045FHM-L00-1S00_BOE_4.5_LCM_overview_36174.html

Generated by mdss-panel-generator

Signed-off-by: Paul Sajna <hello@paulsajna.com>

sajattack committed:
sdm660_defconfig: enable options needed for BlackBerry KEY2 peripherals

Signed-off-by: Paul Sajna <hello@paulsajna.com>

minlexx commented:
Please open CI (actions) checks named "**Device tree validation**", expand the "**Analyze results**" section and fix all the dt-schema validation errors that are fixable within your `.dts`. Some errors come from included `.dtsi` files; those are not your obligation to fix.

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,820 @@
+
+/ {
+	model = "BlackBerry KEY2 LE";
+	compatible = "bbry,luna", "qcom,sdm660", "qcom,sdm636";
```
AFAIK this is incorrect, binding requires only 2 compatibles listed here

sajattack replied:
It said it was unable to match sdm636. But ie whyred only has 636. Should it be 636 or 660?

minlexx replied:
What it says is that your device is missing in https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/arm/qcom.yaml (look at the history of the file)

minlexx replied:
And, perhaps, also Blackberry is missing in https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/vendor-prefixes.yaml

minlexx replied:
As I see it, commits here should be in this order:

* new binding document for panel driver
* panel driver
* vendor-perfixes: add blackberry
* arm/qcom.yaml: add bbry,luna to the list of possible options under qcom,sdm636
* bbry,luna device tree

minlexx replied:
```
sdm636-bbry-luna.dtb: /soc@0/i2c@c175000/port-expander@40: failed to match any schema with compatible: ['st,stmpe1801']
```
Seems this needs to be documented as well

sajattack committed:
dt-bindings: panel: Add boe,bv045fhm-l00

Signed-off-by: Paul Sajna <hello@paulsajna.com>

sajattack committed:
dt-bindings: vendor-prefixes: add BlackBerry Limited as bbry

That's what they referred to themselves as in downstream kernels
(at least TCL ones)

Signed-off-by: Paul Sajna <hello@paulsajna.com>

sajattack committed:
dt-bindings: arm: qcom: New Devices: bbry,luna-boe and bbry,athena-boe

Add bbry,luna-boe under supported sdm636 devices
Add bbry,athena-boe under supported sdm660 devices

Signed-off-by: Paul Sajna <hello@paulsajna.com>

sajattack committed:
dt-bindings: mfd: Add stmpe1801

The mfd is already supported upstream, but not in the dt-bindings.

Signed-off-by: Paul Sajna <hello@paulsajna.com>

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,820 @@
+	gpio_keys {
+			status = "okay";
+			compatible = "gpio-keys";
+			input-name = "gpio-keys";
```
```
sdm636-bbry-luna.dtb: gpio_keys (gpio-keys): 'input-name' does not match any of the regexes: '^(button|event|key|switch|(button|event|key|switch)-[a-z0-9-]+|[a-z0-9-]+-(button|event|key|switch))$', '^pinctrl-[0-9]+$'
```
This likely means this property is unknown and unsupported (remove it? should work without it)

minlexx commented:
All commits that are intended to go upstream should have proper descriptions in commit messages. (Commit modifying sdm660_defconfig, which is not upstream, does not qualify here).

I not fully understand, you have have changed ~1300 lines in this PR, but can't type few words to describe your changes?.. For inspiration l'd suggest to look for similar commits around neighbouring files histories, or ask more questions

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+		i2c1 = &blsp_i2c1;
+		i2c2 = &blsp_i2c2;
+		i2c4 = &blsp_i2c4;
+		i2c6 = &blsp_i2c6;
```
Where are these aliases used?

sajattack replied:
They're not used within the device tree, I was using them so that the devices under `/dev/` would be matched up with the hardware numbering, otherwise I was seeing `/dev/i2c0` `dev/i2c1` `/dev/i2c2` `/dev/i2c3` etc.

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+
+	qcom,msm-id = <0x159 0x0>;
+	qcom,board-id = <0xaa000008 0x00>;
+	qcom,pmic-id = <0x1001b 0x101011a 0x00 0x00 0x1001b 0x201011a 0x00 0x00>;
```
Likely ll of these `qcom,*-id` will have to be removed before sending this upstream.

sajattack replied:
I think it's best to keep these as long as possible.

minlexx replied:
Alright, but these properties are officially deprecated: 
* qcom,msm-id: https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/arm/qcom.yaml#L1174
* qcom,board-id: https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/arm/qcom.yaml#L1213

They are explicitly allowed on some "older" SoCs, https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/arm/qcom.yaml#L1225 the list which includes sdm630 and sdm636, but for some reason not sdm660. That's why pprobably it passes the checks here and not for sdm660-xiaomi-clover :sob:

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+				alloc-ranges = <0x00 0x80000000 0x00 0x20000000>;
+				reusable;
+				alignment = <0x00 0x400000>;
+				size = <0x00 0x800000>;
```
double indentation?

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+				alloc-ranges = <0x00 0x00 0x00 0xffffffff>;
+				reusable;
+				alignment = <0x00 0x400000>;
+				size = <0x00 0x1c00000>;
```
what's wrong with tabs in this block?

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+			qcom,client-id = <1>;
+			qcom,vmid = <QCOM_SCM_VMID_MSS_MSA>;
+		};
+
```
extra newline

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+				reusable;
+				alignment = <0x00 0x400000>;
+				size = <0x00 0x2c00000>;
+				linux,cma-default;
```
extra tabs

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+			compatible = "simple-battery";
+			charge-full-design-microamp-hours = <3000000>;
+			voltage-min-design-microvolt = <3400000>;
+			voltage-max-design-microvolt = <4350000>;
```
extra tabs again

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+	status = "okay";
+
+	touchscreen@38 {
+		compatible = "focaltech,ft8719"; // actually an ft8707 but that's not in mainline
```
C++-style comment is not favoured

This better be just `/* Focaltech ft8707 */` above `touchscreen@38` line

sajattack replied:
changed comment style

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+};
+
+&mdss_dsi0 {
+	status = "okay";
```
status should be last property (before subnodes) with empty line before

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+
+&remoteproc_mss {
+	firmware-name = "qcom/sdm636/luna/mba.mbn", "qcom/sdm636/luna/modem.mbn";
+	status = "okay";
```
status should be last property (before subnodes) with empty line before

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+			regulator-max-microvolt = <3300000>;
+			regulator-enable-ramp-delay = <250>;
+			regulator-allow-set-load;
+			regulator-always-on;
```
Are you sure your device needs this? Usually l3b is always-on only on xiaomi-sdm660 platform.

If you bindly copied regulators block from one of xiaomi devices I suggest you carefully re-check each regulator again.

sajattack replied:
How should I check them without schematics? Just turn them off one by one and see if it still works?

minlexx replied:
From downstream dts if you have it? (`adb pull /sys/firmware/fdt`)

If you don't have downstream dts then just remove regulator-always-on here and see if it works;

The change for xiaomi was originally done in 301cf4f59acf0798dc7700284739e6d4eb1a65ac

minlexx replied:
Were you able to confirm regulators settings?

sajattack replied:
not yet

sajattack replied:
made regulator settings match downstream more closely

minlexx replied:
It seems your regulators setup is different to what we have on xiaomis, e.g. qusbphy is powered by l9a, not l10a?

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,816 @@
+
+&venus {
+	firmware-name = "qcom/sdm636/luna/venus.mbn";
+	status = "okay";
```
empty line before status

sajattack replied:
fixed all whitespace and okay issues mentioned

minlexx commented on drivers/gpu/drm/panel/panel-livata.c:
```diff
@@ -0,0 +1,277 @@
+	struct device *dev = &dsi->dev;
+	int ret;
+
+	mipi_dsi_generic_write_seq(dsi, 0x00, 0x00);
```
geerate usin fresh generator, these should use new multi-functions

sajattack replied:
regenerated

minlexx commented:
Can you attach dmesg of booting to OP post?

sajattack commented:
> Can you attach dmesg of booting to OP post?

added.

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,761 @@
+			keypad,num-columns = <8>;
+
+			linux,keymap = <
+				// Row 1
```
C-style comments seem to be not forbidden, at least they are not mentioned in [dts-coding-style](https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/dts-coding-style.rst), but all examples there use C-style `/* .. */` multiline comments everywhere. This is requirement for linux kernel C source files, not for DTS, but still...  though [devicetree-specification](https://github.com/devicetree-org/devicetree-specification/releases/tag/v0.4) explicitly states 
```
C style (/* ... \*/) and C++ style (//) comments are supported.
```
Perhaps better to change them all to C style comments to be safe, but not a blocker here.

minlexx reviewed (approved):
Alright, I have not checked very thoroughly dts compliance with [dts-coding-style](https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/Documentation/devicetree/bindings/dts-coding-style.rst) - things like order of nodes, order of properties, but I hope at least most style issues are now fixed. Surely kernel maintainers can find more issues, but I don't want to spend more time on this :)

I'll not review keypad patch or panel driver too much - outsdide of my knowledge.
But for panel driver though I think compatible string may be a problem, it should include panel vendor name (wtf is `luna` vendor in `luna,livata`? Is it in `vendor-prefixes.yaml`? no) and perhaps controller name. Is it possible to get information on what panel is actually used there, from some schematics/repair manuals, or by disassembling and looking for part labels? 
I'd suggest to send stmpe pad fixes upstream, and perhaps panel driver too, to get some feedback on those.

sajattack commented:
>  But for panel driver though I think compatible string may be a problem, it should include panel vendor name (wtf is `luna` vendor in `luna,livata`? Is it in `vendor-prefixes.yaml`? no) and perhaps controller name. Is it possible to get information on what panel is actually used there, from some schematics/repair manuals, or by disassembling and looking for part labels?

For panel name I suppose I had a misunderstanding. I thought tianma in tianma,td4310 referred to xiaomi-beryllium-tianma, but it turns out tianma is a company name. I was thinking maybe for an unknown panel it was ok to use the name of the device and the name of the panel. I'll try to see if I can find more info and if not maybe change it to bbry,livata.

sajattack commented:
> > But for panel driver though I think compatible string may be a problem, it should include panel vendor name (wtf is `luna` vendor in `luna,livata`? Is it in `vendor-prefixes.yaml`? no) and perhaps controller name. Is it possible to get information on what panel is actually used there, from some schematics/repair manuals, or by disassembling and looking for part labels?
> 
> For panel name I suppose I had a misunderstanding. I thought tianma in tianma,td4310 referred to xiaomi-beryllium-tianma, but it turns out tianma is a company name. I was thinking maybe for an unknown panel it was ok to use the name of the device and the name of the panel. I'll try to see if I can find more info and if not maybe change it to bbry,livata.

all I know about it is this dtsi from downstream https://github.com/tim-ecoder/blackberry-android-linux-kernel/blob/sdm660-sdm636/ACW142/krab-luna-droidian/arch/arm64/boot/dts/qcom/dsi-panel-livata-video.dtsi

sajattack commented:
> > > But for panel driver though I think compatible string may be a problem, it should include panel vendor name (wtf is `luna` vendor in `luna,livata`? Is it in `vendor-prefixes.yaml`? no) and perhaps controller name. Is it possible to get information on what panel is actually used there, from some schematics/repair manuals, or by disassembling and looking for part labels?
> > 
> > 
> > For panel name I suppose I had a misunderstanding. I thought tianma in tianma,td4310 referred to xiaomi-beryllium-tianma, but it turns out tianma is a company name. I was thinking maybe for an unknown panel it was ok to use the name of the device and the name of the panel. I'll try to see if I can find more info and if not maybe change it to bbry,livata.
> 
> all I know about it is this dtsi from downstream https://github.com/tim-ecoder/blackberry-android-linux-kernel/blob/sdm660-sdm636/ACW142/krab-luna-droidian/arch/arm64/boot/dts/qcom/dsi-panel-livata-video.dtsi

I think it's boe. https://www.alibaba.com/product-detail/BOE-BV045FHM-L00-1S00-4-5_1600095821720.html 
There's a comment in the dtsi that mentions boe and all the replacement screens you see for sale online say "b4502e1f" on the fpc.

minlexx commented on drivers/gpu/drm/panel/Kconfig:
```diff
@@ -385,6 +385,17 @@ config DRM_PANEL_LINCOLNTECH_LCD197
+	  The panel has a 1080x1620 resolution and uses
+	  24 bit RGB per pixel. It provides a MIPI DSI interface to
+	  the host.
+
```
I'mafraid this block will have to be reordered to alphabetically be around other BOE panels.

Same for Makefile (and I assume sdm660_defconfig will be changed too)

minlexx commented on arch/arm64/boot/dts/qcom/sdm636-bbry-luna.dts:
```diff
@@ -0,0 +1,43 @@
+/dts-v1/;
+
+#include "sdm636.dtsi"
+#include "sdm660-bbry-common.dtsi"
```
I think first should come include `-common`, then `sdm636`

Because common includes `sdm660.dtsi`, and `sdm636.dtsi` also includes `sdm660.dtsi` and then applies overrides on top

minlexx replied:
Though maybe it doesn't matter, because https://github.com/sdm660-mainline/linux/blob/qcom-sdm660-6.15.y/arch/arm64/boot/dts/qcom/sdm636.dtsi so far only overrides compatible string on `&adreno_gpu` and dtc combines them all together while compiling anyways :thinking:

sajattack replied:
oop, compiler didn't like that

sajattack replied:
I think maybe I should pull 660 out of common?

minlexx replied:
alright, whatever works..

minlexx commented on arch/arm64/boot/dts/qcom/sdm660-bbry-athena.dts:
```diff
@@ -0,0 +1,42 @@
+/dts-v1/;
+
+#include "sdm660-bbry-common.dtsi"
+/ {
```
Add empty line above / after include line

sajattack commented:
pushing some changes for panels I haven't tested yet, converting to draft.

minlexx commented on arch/arm64/boot/dts/qcom/sdm660-bbry-common.dtsi:
```diff
@@ -0,0 +1,756 @@
+
+/delete-node/ &qhee_code;
+/delete-node/ &tz_mem;
+/delete-node/ &buffer_mem;
```
This buffer_mem is no longer there, it was renamed to cdsp_region for sdm660 and at the same time also deleted in sdm636.dtsi ( https://github.com/sdm660-mainline/linux/commit/13b651e1389a6301df5debedc96c408761df2aa1)

sajattack commented:
> pushing some changes for panels I haven't tested yet, converting to draft.

couldn't get the new variant to work, and don't want it to block this any longer, removing synaptic variant to unblock.

r00t- commented:
@sajattack :
> > pushing some changes for panels I haven't tested yet, converting to draft.
> 
> couldn't get the new variant to work, and don't want it to block this any longer, removing synaptic variant to unblock.

maybe push those commits somewhere else for further testing?

edit: https://github.com/sajattack/sdm660-linux/tree/saj/bbry/syna-panels

eMPee584 commented:
**TLDR**: used Key2 (6/64GB) on kleinanzeigen.DE 🇩🇪 for just below 80€:
https://www.kleinanzeigen.de/s-anzeige/blackberry-key2-scheibe-gebrochen-funktional/3236520191-173-16744
left upper screen corner is cracked (that could be improved a lot with some thinned clear nail polish, i.e. letting it dry and scraping residual off with a cutter blade). I also got a spare display foil and rubber case which I can send by mail.

<img width="764" height="728" alt="Screenshot 2025-11-14 at 12-31-17 Blackberry Key2 Scheibe gebrochen funktional in Düsseldorf - Gerresheim Telefon gebraucht kaufen kleinanzeigen de" src="https://github.com/user-attachments/assets/7093767d-d301-4071-be43-f0775dfbed05" />

----

Hi guys and thanks for the work on this great device!
I've got a key2LE and as my brand new titan2 arrived a week ago, I'm planning to use that for android and put PMOS on the blackberry, hopefully being able to contribute one or the other patch on the user stack as time admits.
As summarized above, I spotted a key2 (6gb RAM!) in good functional condition for a bargain price. I would love to get it myself, but realistically, I'm short both on cash and time to spend on debugging kernel stuff the coming months as I'm quite busy preparing for my master electrician exams early next year.
So maybe you know a fellow kernel dev who would have fun bringing this one to work better.  @sajattack maybe you ~~can also fwd it to the PMOS list or~~ know someone from Germany who could take advantage of this.
Best Regards from Dresden 🇪🇺🙃
